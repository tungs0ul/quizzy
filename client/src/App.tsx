import { useState, useEffect } from "react";
import useWebSocket from "react-use-websocket";
import useLocalStorage from "./hooks/useLocalStorage";
import StudentView from "./components/views/StudentView";
import AdminView from "./components/views/AdminView";
import { BrowserRouter, Routes, Route, Link } from "react-router-dom";

// const WSURI = "ws://localhost:8080/ws";
const WSURI =
  ((window.location.protocol == "https:" && "wss://") || "ws://") +
  window.location.host +
  "/ws";

const initQuestion = {
  msg: "question",
  data: "",
  options: [],
};

export type Question = {
  type: string;
  data: string;
  options: string[];
};

export type Result = {
  name: string;
  value: number;
};

function App() {
  const { sendMessage, lastMessage } = useWebSocket(WSURI, {
    shouldReconnect: () => true,
    reconnectInterval: 5000,
  });
  const [question, setQuestion] = useLocalStorage("question", initQuestion);
  const [answer, setAnswer] = useLocalStorage("answer", null);
  const [result, setResult] = useState<Result[]>([]);
  const [notice, setNotice] = useState("");
  const [active, setActive] = useState(false);
  const [intervalID, setIntervalID] = useState<number | null>(null);

  useEffect(() => {
    if (lastMessage !== null) {
      if (!lastMessage.data) {
        return;
      }
      const data = JSON.parse(lastMessage.data);
      if (data.type === "question") {
        if (data.data !== question.data) {
          setNotice("The question has been changed!");
          setAnswer(null);
          setQuestion(data);
          setResult([]);
          if (intervalID !== null) {
            clearInterval(intervalID);
          }
          let id = setInterval(() => setNotice(""), 1500);
          setIntervalID(id);
        }
      } else if (data.type === "result") {
        if (
          data.data.reduce((acc: number, cur: number) => acc + cur, 0) === 0
        ) {
          setAnswer(null);
        }
        let result: Result[] = [];
        data.data.forEach((e: number, i: number) => {
          result.push({ name: question.options[i], value: e });
        });
        setResult(result);
      } else if (data.type === "clear") {
        setAnswer(null);
        setNotice("The answer has been cleared!");
        setInterval(() => setNotice(""), 1500);
      } else if (data.type === "reset") {
        setQuestion(initQuestion);
        setAnswer(null);
        setResult([]);
      }
    }
  }, [lastMessage]);

  return (
    <BrowserRouter>
      <div className="h-screen relative">
        <div className="flex justify-between items-center pb-4 pr-10 header">
          <div className="cursor-pointer p-4 text-yellow-600 hover:text-green-600 text-4xl items-center h-full flex justify-center">
            <Link to="/login">Admin</Link>
          </div>
          <div className="flex justify-end tablet:text-12xl laptop:text-16xl mobile:text-6xl font-bold">
            QUIZZY
          </div>
        </div>
        <Routes>
          <Route
            path="/"
            element={
              <StudentView
                answer={answer}
                question={question}
                result={result}
                sendMessage={sendMessage}
                setAnswer={setAnswer}
                notice={notice}
              />
            }
          />
          <Route
            path="/login"
            element={
              <AdminView
                sendMessage={sendMessage}
                setResult={setResult}
                result={result}
                question={question}
              />
            }
          />
        </Routes>
      </div>
    </BrowserRouter>
  );
}

export default App;
