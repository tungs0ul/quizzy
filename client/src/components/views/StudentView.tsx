import React, { useCallback, Suspense } from "react";
import { SendMessage } from "react-use-websocket";
import { COLORS } from "../../utils";
import { motion } from "framer-motion";
import { Question, Result } from "../../App";

const PieGraph = React.lazy(() => import("../graph/PieGraph"));

type Props = {
  answer: number | null;
  setAnswer: any;
  sendMessage: SendMessage;
  question: Question;
  result: Result[];
  notice: string;
  active: boolean;
};

export default function StudentView({
  answer,
  setAnswer,
  sendMessage,
  question,
  result,
  notice,
  active,
}: Props) {
  const submitAnswer = useCallback(
    (index: number) => {
      sendMessage(
        JSON.stringify({ type: "answer", data: JSON.stringify(index) })
      );
      setAnswer(index);
    },
    [answer]
  );

  return (
    <div>
      {notice && (
        <motion.div
          initial={{ opacity: 0, y: -100 }}
          animate={{ opacity: 1, y: 20 }}
          transition={{ duration: 0.5 }}
          className="fixed right-5 top-0 bg-gray-200 h-14 flex items-center text-center z-50"
        >
          <div className="p-2 text-2xl">{notice}</div>
          <motion.div
            initial={{ width: 0 }}
            animate={{ width: "100%" }}
            transition={{ duration: 1.5 }}
            className="bg-blue-600 w-full absolute bottom-0 h-2"
          ></motion.div>
        </motion.div>
      )}
      {answer === null && active ? (
        <div className="relative">
          {question && (
            <>
              <div className="flex justify-center text-4xl p-4 tablet:pt-12 tablet:pb-12 mobile:pt-4 mobile:pb-8 mobile:mb-8 tablet:mb-1 relative bubble">
                <p className="thought">{question.data}</p>
              </div>
              <div className="gap-4 flex flex-col mobile:flex-col laptop:flex-row justify-center tablet:flex-wrap p-8 options">
                {question?.options &&
                  question.options.map((option: string, index: number) => (
                    <div
                      className="flex justify-center"
                      key={`${option}-${index}`}
                    >
                      <motion.span
                        initial={{ opacity: 0, width: 0 }}
                        animate={{
                          opacity: 1,
                          x: 0,
                          width: "100%",
                        }}
                        transition={{ duration: index * 0.5 + 0.5 }}
                        onClick={() => submitAnswer(index)}
                        className="border-2 border-black mobile:p-4 tablet:p-12 mobile:w-full flex justify-center rounded-md mobile:text-3xl tablet:text-4xl cursor-pointer option"
                      >
                        {option}
                      </motion.span>
                    </div>
                  ))}
              </div>
            </>
          )}
        </div>
      ) : (
        <div>
          <div className="flex justify-center text-4xl p-4 tablet:pt-12 tablet:pb-12 mobile:pt-4 mobile:pb-8 mobile:mb-8 tablet:mb-1 relative bubble">
            <p className="thought">{question?.data?.length && question.data}</p>
          </div>
          {answer !== null && (
            <div className="flex justify-center text-center text-2xl">
              Your answer is:{" "}
              <span className="ml-2" style={{ color: COLORS[answer] }}>
                {question?.options[answer]}
              </span>
            </div>
          )}
          <Suspense fallback={<div>Loading...</div>}>
            <PieGraph data={result} />
          </Suspense>
        </div>
      )}
    </div>
  );
}
