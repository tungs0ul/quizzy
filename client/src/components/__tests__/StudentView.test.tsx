import TestRenderer, { act } from "react-test-renderer";
import { Result, Question } from "../../App";
import * as React from "react";
import StudentView from "../views/StudentView";

const answer = null;
const setAnswer = jest.fn();
const sendMessage = jest.fn();
const question: Question = {
  type: "question",
  data: "Yes Or No",
  options: ["Yes", "No"],
  uuid: "1",
};
const result: Result[] = [];
const notice = "";
const active = true;

it("render student view without notice", () => {
  const test = TestRenderer.create(
    <StudentView
      answer={answer}
      setAnswer={setAnswer}
      question={question}
      active={active}
      notice={notice}
      sendMessage={sendMessage}
      result={result}
    />
  );
  const x = test.root.findAllByType("span")[0];
  x.props.onClick();
  expect(test.toJSON()).toMatchSnapshot();
});

it("render student view with notice", () => {
  const test2 = TestRenderer.create(
    <StudentView
      answer={answer}
      setAnswer={setAnswer}
      question={question}
      active={active}
      notice={"hello"}
      sendMessage={sendMessage}
      result={result}
    />
  );

  expect(test2.toJSON()).toMatchSnapshot();
});
