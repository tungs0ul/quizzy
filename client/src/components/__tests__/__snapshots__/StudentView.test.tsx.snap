// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render student view with notice 1`] = `
<div>
  <div
    className="fixed right-5 top-0 bg-gray-200 h-14 flex items-center text-center z-50"
    style={
      Object {
        "opacity": 0,
        "transform": "translateY(-100px) translateZ(0)",
      }
    }
  >
    <div
      className="p-2 text-2xl"
    >
      hello
    </div>
    <div
      className="bg-blue-600 w-full absolute bottom-0 h-2"
      style={
        Object {
          "width": "0px",
        }
      }
    />
  </div>
  <div
    className="relative"
  >
    <div
      className="flex justify-center text-4xl p-4 tablet:pt-12 tablet:pb-12 mobile:pt-4 mobile:pb-8 mobile:mb-8 tablet:mb-1 relative bubble"
    >
      <p
        className="thought"
      >
        Yes Or No
      </p>
    </div>
    <div
      className="gap-4 flex flex-col mobile:flex-col laptop:flex-row justify-center tablet:flex-wrap p-8 options"
    >
      <div
        className="flex justify-center"
      >
        <span
          className="border-2 border-black mobile:p-4 tablet:p-12 mobile:w-full flex justify-center rounded-md mobile:text-3xl tablet:text-4xl cursor-pointer option"
          onClick={[Function]}
          style={
            Object {
              "opacity": 0,
              "width": "0px",
            }
          }
        >
          Yes
        </span>
      </div>
      <div
        className="flex justify-center"
      >
        <span
          className="border-2 border-black mobile:p-4 tablet:p-12 mobile:w-full flex justify-center rounded-md mobile:text-3xl tablet:text-4xl cursor-pointer option"
          onClick={[Function]}
          style={
            Object {
              "opacity": 0,
              "width": "0px",
            }
          }
        >
          No
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`render student view without notice 1`] = `
<div>
  
  <div
    className="relative"
  >
    <div
      className="flex justify-center text-4xl p-4 tablet:pt-12 tablet:pb-12 mobile:pt-4 mobile:pb-8 mobile:mb-8 tablet:mb-1 relative bubble"
    >
      <p
        className="thought"
      >
        Yes Or No
      </p>
    </div>
    <div
      className="gap-4 flex flex-col mobile:flex-col laptop:flex-row justify-center tablet:flex-wrap p-8 options"
    >
      <div
        className="flex justify-center"
      >
        <span
          className="border-2 border-black mobile:p-4 tablet:p-12 mobile:w-full flex justify-center rounded-md mobile:text-3xl tablet:text-4xl cursor-pointer option"
          onClick={[Function]}
          style={
            Object {
              "opacity": 0,
              "width": "0px",
            }
          }
        >
          Yes
        </span>
      </div>
      <div
        className="flex justify-center"
      >
        <span
          className="border-2 border-black mobile:p-4 tablet:p-12 mobile:w-full flex justify-center rounded-md mobile:text-3xl tablet:text-4xl cursor-pointer option"
          onClick={[Function]}
          style={
            Object {
              "opacity": 0,
              "width": "0px",
            }
          }
        >
          No
        </span>
      </div>
    </div>
  </div>
</div>
`;
